lib/base/context.hpp

```diff
@@ -33,18 +34,21 @@ std::ostream& operator<<(std::ostream& stream, const ContextTrace& trace);
 class ContextFrame
 {
 public:
-	ContextFrame(const String& message);
+	ContextFrame(std::function<void(std::ostream&)> message);
 	~ContextFrame();
 
 private:
-	static std::list<String>& GetFrames();
+	static std::vector<std::function<void(std::ostream&)>>& GetFrames();
 
 	friend class ContextTrace;
 };
 
 /* The currentContextFrame variable has to be volatile in order to prevent
  * the compiler from optimizing it away. */
-#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame(message)
+#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame ([&](std::ostream& stream) { \
+stream << message; \
+})
```

Comment:
This change means that the caller has to ensure that whatever expression is used leads to a constant result while the scope is active. Not impossible but requires review of all callers.